From 6a251ded10aa51735ed14a5b3220a2d8895ab268 Mon Sep 17 00:00:00 2001
From: Naganarasimha <naganarasimha_gr@apache.org>
Date: Mon, 23 Oct 2017 23:41:38 +0530
Subject: [PATCH 2762/2848] HADOOP-14966. Handle JDK-8071638 for
 hadoop-common. Contributed by Bibin A Chundatt.

(cherry picked from commit 6b8122458ede9ececb138bed99d368580e5439a1)
(cherry picked from commit 5ee67c12025e36142572d50ce18b3a32722497d3)

Change-Id: I328c919c214bc5af84b938ecc3bd85795f1a746b
---
 .../hadoop/util/concurrent/ExecutorHelper.java     |   10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/concurrent/ExecutorHelper.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/concurrent/ExecutorHelper.java
index 3bc9ed9..a617558 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/concurrent/ExecutorHelper.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/concurrent/ExecutorHelper.java
@@ -40,14 +40,16 @@ static void logThrowableFromAfterExecute(Runnable r, Throwable t) {
 
     //For additional information, see: https://docs.oracle
     // .com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor
-    // .html#afterExecute(java.lang.Runnable,%20java.lang.Throwable) .
+    // .html#afterExecute(java.lang.Runnable,%20java.lang.Throwable)
 
-    if (t == null && r instanceof Future<?>) {
+    // Handle JDK-8071638
+    if (t == null && r instanceof Future<?> && ((Future<?>) r).isDone()) {
       try {
         ((Future<?>) r).get();
       } catch (ExecutionException ee) {
-        LOG.warn("Execution exception when running task in " +
-            Thread.currentThread().getName());
+        LOG.warn(
+            "Execution exception when running task in " + Thread.currentThread()
+                .getName());
         t = ee.getCause();
       } catch (InterruptedException ie) {
         LOG.warn("Thread (" + Thread.currentThread() + ") interrupted: ", ie);
-- 
1.7.9.5

